### ISSUE TITLE
Corrupt .leo files cause Leo to hang

### ISSUE THREAD

**Issue by edreamleo** [leo-editor/leo-editor#158]
As reported by vitalije 	

this morning I found one of my Leo files,  that I worked on a few days ago, corrupted. While opening the file, Leo hanged and I had to kill it. After some examination of xml source I found that some of my last added nodes got the same gnx. One of them had child with the same gnx and that caused Leo to hang. I have changed manually gnx of the parent node in xml. When I finally open outline in Leo I found few other nodes were missing or had wrong headline/body.
I don't know what caused this and can't reproduce it but I strongly suspect it has something to do with recent changes in Leo write.
Unfortunately I can't say which revision I used to write the problematic file. It was one of the revisions after 934537266cf5d3e556025adc6ba39ff35be81741 which solved bug #138. This morning I pulled latest Leo revision before trying to open file.

I am using Ubuntu 14.04 64bit,
Leo 5.0-final, build 20150305163031, Thu Mar  5 16:30:31 CST 2015
Git repo info: branch = master, commit = 4dddcc1ac5d8
Python 2.7.6, PyQt version 4.8.6
linux2

**Comment by edreamleo** [comment]
Here are notes from my initial reply:
​
​> I strongly suspect it has something to do with recent changes in Leo write.
The write code doesn't allocate gnx's. Leo uses a "holding" logic during reads. Thereafter, gnx's are allocated when a vnode is created. The culprit is more likely to be elsewhere:

- Bug 130:  About rev c6c14ef on March 2. This changes how Leo calculates the smallest "safe" gnx to use for new nodes.  https://github.com/leo-editor/leo-editor/issues/130.  It's my prime suspect.

- Rev 4ac015f, March 10, removes allow_cloned_sibs logic.
- Rev 4cf0945 and following, March 10: remove at.writeVersion5.
- Bug 149: Rev 6439440, March 11. Changes how Leo reports recovered nodes.

**Important**: Iirc, there was a *brief* window, perhaps an hour, perhaps less, in which Leo may have allocated duplicate gnx's.  This was due to a blunder fixing #130.  It's possible that this bug could be the result, but that's far from an adequate response.

The initial fix will be to add code to Leo's .leo read logic.  This new code will check for duplicate gnx's as the .leo file is being read. The proper place to do this may be p._linkAsNthChild.  If the to-be-linked node has the same gnx as one of its ancestors, a warning will be given and the node will get a new gnx.

A good test of this logic will be to load the corrupted file that vitalije sent me.  I expect to add this code later today.

**Comment by edreamleo** [comment]
Rev 60d61e1 adds a *hack* that allows Leo to read at least some of the corrupt .leo files without looping.

I'm not convinced that the real problem involves gnx's.  Perhaps the @clean write code sometimes writes the root @clean node as its own child. That would cause Leo to hang as reported, even if the gnx-allocation logic were sound.  Indeed, the essence of the corrupt .leo file is::

    <v t="vitalije.20150316130845.1"><vh>@clean stylesheets/main.less</vh>
    <v t="vitalije.20150316130845.1"></v>
    ...
    </v>

As you can see, the first child of the @clean node has the same gnx as the root node.  Moreover, this first child has no `<vh>` element which is unexpected.  This increases suspicion of the @clean write logic.

**Comment by edreamleo** [comment]
Here is the checkin log for Rev 96f8bc4:

The --check-outline command-line switch is now functional. OFF by default: it's expensive.
All tests pass regardless of the switch.
**Warning**: the gnx checks in c.checkOutline will hang if nodes are actually their own ancestors.
I'll add one or more safe generators asap.

**Comment by edreamleo** [comment]
Rev a7d764a:

Added c.safe_all_positions and its helper, p.safeMoveToThreadNext.
p.safeMoveToThreadNext just calls g.error and returns p.moveToParent on serious structure errors.
c.checkGnxs (called by c.checkOutline) now calls c.safe_all_positions, so it shouldn't hang before it completes.  However, Leo will likely still hang later.

**Next**: p.safeMoveToThreadNext will allocate a new vnode when reporting a structure error and will allocate a new gnx when reporting a gnx error.  This should prevent Leo from hanging after c.checkGnxs reports an error.

**Comment by edreamleo** [comment]
Rev ada9a0b:

Improved error correction on serious outline errors.
fc.createSaxVnode corrects the error only when --check-outline is not in effect.
p.safeMoveToThreadNext now simply *deletes* vnodes that are their own parents.
This strategy works with vitalije's erroneous file.

In short, --check-outline should be fairly safe for vitalije to use.

**Comment by edreamleo** [comment]
A **structurally-corrupted outline** is an outline in which a vnode is an ancestor of itself.
A **gnx-corrupted outline** is an outline in which two *distinct* vnodes have the same gnx.
A **corrupted .leo file** is a .leo file for which two distinct <v> elements have the same gnx.
Corrupt .leo files *always* cause the outline to become structurally corrupted when Leo reads it later.

Leo will almost certainly hang if the outline becomes structurally corrupted.  Afaik, this has never happened to vitalije while he is working with Leo.  Rather, a corrupt .leo file caused the outline to be structurally corrupted when Leo read the .leo file later. 

With this in mind, here are the possible culprits for this bug:
1. [Early corruption] Leo's data structures became gnx-corrupted sometime before writing the @clean tree.  The possible causes for this gnx corruption:
- An error in the @clean *read* logic.
- An error in the outline (sax) *read*.
- An error in the paste or paste-retaining clones logic.
2. [Late corruption] Some part of the @clean *write* logic mistakenly the .leo file.

The --check-outline checks should be able to detect all forms of early corruption.  If no such problems are reported, we can start to suspect late corruption more strongly.

**Comment by edreamleo** [comment]
**About error recovery in p.safeMoveToThreadNext**
At present, this method simply *deletes* nodes when the outline is seen to be structurally-corrupted.

Whoa!  Won't this lose data? Actually not.  There are two cases:

1. vitalije has never reported that Leo hangs while he was working on an outline. Therefore, p.safeMoveToThreadNext will not destroy data in such cases.  p.safeMoveToThreadNext might reassign gnx's, but that can't destroy data.

2. Leo only hangs when reading a .leo file.  There are <v> elements with the same gnx, but both share the same <t> element, so *the data was lost when Leo wrote the file*.

This is good news.  There is no need for p.safeMoveToThreadNext to be "careful" about recovery from structural corruption.  There is no need to pass flags from c.safe_all_positions to p.safeMoveToThreadNext.

**Comment by edreamleo** [comment]
Rev 5d482c8: **Settings to use instead of --check-outline**

Using these settings may be more convenient than using --check-outline::

    @bool check_outline_after_read = False
    @bool check_outline_before_save = False

The defaults are as shown.  Setting them both to True will partially take the place of --check-outline.
**Note**: These settings have existed for a long time, but I have only just placed them in leoSettings.leo.

**Comment by edreamleo** [comment]
**Major bugs fixed in fc.createSaxVnode**
Rev f14a277 fixes three major problems in createSaxVnode:

1. The test ``if sax_node.tnx:`` is wrong.  It must be: ``if sax_node.tnx is not None:``.
2. Leo must *never* call ``ni.getNewIndex(None)``!  Instead, createSaxVnode now passes ``gnx=None`` to the VNode ctor which then calls ``ni.getNewIndex(v=self)``.
3. Similarly, the "recovery" code in createSaxVnode must not pass a gnx to the VNode ctor.

These bugs were revealed when Ross Burnett reported this crash::

    File ...\leo\core\leoNodes.py", line 88, in end_holding
        if v.fileIndex:
    AttributeError: 'NoneType' object has no attribute 'fileIndex'
    
As a double check, ni.getNewIndex now calls g.internalError if v is None.

I am going to leave this bug open for at least another week, but this is an important step forward.

**Comment by edreamleo** [comment]
It appears that I have been bitten by this bug myself.  On two occasions LeoDocs.leo has apparently lost data due to two nodes unexpectedly becoming clones under highly suspicious circumstances.  As a result, I have added the following two settings to LeoDocs.leo::

    @bool check_outline_after_read = True
    @bool check_outline_before_save = True

This instantly revealed a minor flaw in check-outline.  It will complain about empty v.fileIndex for all nodes in an @auto file.  I'll fix this asap.

**Comment by vitalije** [comment]
I have just updated Leo and while opening one of files that I have edited this morning, I have following message in Log pane
```
Leo Log Window
Leo 5.0-final, build 20150316085617, Mon Mar 16 08:56:17 CDT 2015
Git repo info: branch = master, commit = d300caedef6e
Python 2.7.6, PyQt version 4.8.6
linux2
reading: /home/vitalije/play/probe/kviziracija/kviziracija.leo
createSaxVnode: duplicate parent gnx vitalije.20150318103114.2 
 <VNode 7f358c032a50 podaci katastar>
read 23 files in 0.26 seconds
read outline in 0.43 seconds
```

**Comment by vitalije** [comment]
I have activated checks after read and before save and opened corrupted outline to fix errors.
I lost content of two nodes. One of them was read as clone of some other node and that way it lost original content. The other one had the same content as one of its children, but they were not marked as clones (otherwise they would make circular path).

Then, I have deleted one of wrongly cloned nodes and inserted new node instead. Then I wrote body that was lost. 
When I tried to save corrected Leo file, I got following error:
```
multiple vnode with same gnx: vitalije.20150318152029.1 vnodes: [[<VNode 7fa352965b90 copyjs: podaci katastra>, <VNode 7fa37b433850 podaci>]]
1 nodes checked 1 errors
outline not written
```

Then I cut newly written node, and pasted it again twice. Then I deleted original node and after that Leo correctly saved file.

HTH

**Comment by edreamleo** [comment]
When I tried to save corrected Leo file, I got following error::

    multiple vnode with same gnx: [snip]
    1 nodes checked 1 errors
    outline not written

I got the same error today.  Leo should simply reallocate the gnx's and allow the file to be saved.  I'll fix this in the next hour or so.

Imo, the recent problems with duplicate gnx's arose from the bug that was fixed early this morning.  This bug had the effect of assigning the *empty string* to vnodes in some situations. I think that situation is gone for good, but we may be dealing with the consequences of the bug for some time...

**Comment by edreamleo** [comment]
Rev d006eb7 rewrites the check-outline logic.  It now re-allocates gnx's if the gnx is ``''`` or a duplicate. We'll see how well this works if and when it is needed.

c.checkOutline now returns the number of structural errors. fc.write_Leo_file will refuse to write the .leo file only if structural errors exist. Afaik, this has never happened while writing a .leo file: it is the *read* logic that created the structural errors in the past as the result of duplicate gnx's.

In short, reallocating gnx's as needed should correct the effects of previous bugs, and that may be the end of the matter. We shall see...

**Comment by edreamleo** [comment]
A few more tweaks to the checking code:

- c.checkGnxs now prints more detailed data to the console, and just an "OK" message to the log. Both console and log get the same detailed message in case of errors.
- c.checkGnxs now prints *all* error messages in red, not just the summary message.
- c.checkGnxs prints the time taken for the check.  About 0.15 sec for 10,000 nodes. No big deal.
- The check-outline command does full checks, including link checks.  This takes considerably longer.

**Comment by edreamleo** [comment]
I saw an unexpected clone late yesterday.  However, both clones were adjacent siblings and contained identical descendants, so it may have been simply a wayward keystroke.  The automatic check-outline settings were in effect all yesterday. They never complained.

I am reporting this here because as a note to myself to *save* any such unexpected .leo file before simply changing it!  It's important that there be *no* problems reported for at least a week before I close this bug.

**Comment by edreamleo** [comment]
I am seriously considering *always* checking .leo files after reads and before writes:
https://groups.google.com/d/msg/leo-editor/9STvVMI5Ku4/NpmRqkLJU0AJ
This would make Leo safer and would remove several options, including --check-outline.

**Comment by JohnLunzer** [comment]
I'm pretty sure that I updated my installation after your last comment 2 days ago, and I was still getting corrupted clones and losing clones. When I enabled --check-outline the issues seemed to go away. Sorry I can't provide more details.

**Comment by edreamleo** [comment]
On Fri, Mar 20, 2015 at 7:37 PM, JohnLunzer <notifications@github.com>
wrote:

> I'm pretty sure that I updated my installation after your last comment 2
> days ago, and I was still getting corrupted clones and losing clones. When
> I enabled --check-outline the issues seemed to go away. Sorry I can't
> provide more details.
>
​There was a period in which the Leo code could have produce .leo files
that were time bombs: they would cause problems when next opened.  This is
reason enough always to check .leo files before writing them and after
reading them.  It will happen tomorrow.

Edward

**Comment by edreamleo** [comment]
Leo just reported a clone clash after cutting an outline from leoSettings.leo and pasting into leoPy.leo::

    # Only in console.  Should also go to log.
    getLeoOutlineFromClipboard can not happen: index clash ekr.20150320233042.9
    <VNode be2b530 @string rst3_ignore_prefix = @rst-ignore>
    
    # Reported to the log and to the console.
    multiple vnodes with gnx: u'ekr.20150320233042.7'
    new gnx: ekr.20150320233042.7 <VNode b4df190 To be changed>
    new gnx: ekr.20150320233042.7 <VNode be2b3b0 @string rst3_ignore_headlines_prefix = @rst-no-headlines>
    multiple vnodes with gnx: u'ekr.20150320233042.9'
    new gnx: ekr.20150320233042.9 <VNode abf4b50 Options that set command names>
    new gnx: ekr.20150320233042.9 <VNode be2b530 @string rst3_ignore_prefix = @rst-ignore>
    check-outline ERROR! leoPy.leo 8254 nodes, 4 gnx errors, 0 structure error

Let's say this is *good* news.  It's an important clue.  It also highlights the need for permanent gnx checks.

This bug must remain open until I fully understand the source of the gnx clashes.

**Comment by edreamleo** [comment]
I am going to close this bug.  Ripping out the gnx-caching scheme should be a complete fix.
In the unlikely event that problems remain, they would not be as serious because of Leo's new recover logic. As a result, I have also removed the major label.
